pennylane/decomposition/decomposition_graph.py

```diff
@@ -160,29 +168,73 @@ def _recursively_add_op_node(self, op_node: CompressedResourceOp) -> int:
 
         op_node_idx = self._graph.add_node(op_node)
         self._op_node_indices[op_node] = op_node_idx
+
         if op_node.op_type.__name__ in self._target_gate_set:
             self._target_gate_indices.add(op_node_idx)
             return op_node_idx
 
+        if op_node.op_type is qml.ops.Controlled or op_node.op_type is qml.ops.ControlledOp:
+            # This branch only applies to general controlled operators
+            return self._add_controlled_decomp_node(op_node, op_node_idx)
+
         for decomposition in self._get_decompositions(op_node.op_type):
-            d_node_idx = self._recursively_add_decomposition_node(decomposition, op_node.params)
+            decomp_resource = decomposition.compute_resources(**op_node.params)
+            d_node_idx = self._recursively_add_decomposition_node(decomposition, decomp_resource)
             self._graph.add_edge(d_node_idx, op_node_idx, 0)
 
         return op_node_idx
 
-    def _recursively_add_decomposition_node(self, rule: DecompositionRule, params: dict) -> int:
+    def _add_special_decomp_rule_to_op(
+        self, rule: DecompositionRule, op_node: CompressedResourceOp, op_node_idx: int
+    ) -> int:
+        """Adds a special decomposition rule to the graph."""
+        decomp_resource = rule.compute_resources(**op_node.params)
+        d_node_idx = self._recursively_add_decomposition_node(rule, decomp_resource)
+        self._graph.add_edge(d_node_idx, op_node_idx, 0)
+        return op_node_idx
```

Comment:
Is there a scenario in which the returned `op_node_idx` has been modified? :thinking: If not, we should not be returning it. If it is being modified, we should be returning it for the general case in `_add_controlled_decomp_node` below, no?

Reply:
Removed the return statement.